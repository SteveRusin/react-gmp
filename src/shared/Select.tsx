import { TriangleDown } from '@styled-icons/octicons';
import styled, { useTheme } from 'styled-components';

const SelectTriangle = styled(TriangleDown)`
  color: ${(props) => props.color};
`;

const SelectWrapper = styled.div`
  display: flex;
  align-items: center;
  max-width: 150px;

  cursor: pointer;
`;

// todo make it interactive in next HW

export const Select = () => {
  const { palette } = useTheme();
  return (
    <SelectWrapper>
      <span>RELEASE DATE</span>
      <SelectTriangle color={palette.primary} size='20' />
    </SelectWrapper>
  );
};
