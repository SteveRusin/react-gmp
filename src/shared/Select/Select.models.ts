export interface SelectOption<TId> {
  id: TId;
  name: string;
}

export interface SelectProps<TId> {
  options: SelectOption<TId>[];
  value?: TId | null;
  onSelect: (optionId: TId) => void;
  placeholder?: string;
}
