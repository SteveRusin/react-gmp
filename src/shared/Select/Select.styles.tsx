import { darken } from 'polished';

import { TriangleDown } from '@styled-icons/octicons';
import styled from 'styled-components';

export const SelectTriangle = styled(TriangleDown)`
  position: absolute;
  right: 0;
  top: 50%;
  transform: translateY(-50%);
  color: ${(props) => props.color};
`;

export const SelectWrapper = styled.div`
  display: flex;
  align-items: center;
  min-width: 150px;

  cursor: pointer;
  position: relative;

  .select-placeholder {
    opacity: 0.5;
  }

  .placeholder-wrapper {
    position: relative;
    display: flex;
    width: 100%;
  }

  .select-trigger {
    width: 100%;
  }
`;

export const OptionsWrapper = styled.div`
  position: absolute;
  width: 100%;
  left: 0;
  bottom: 0;
  transform: translateY(100%);
  z-index: 1;
  background-color: ${({ theme }) => theme.palette.accent};
  border-radius: 4px;
  box-shadow: 0px 0px 2px rgba(0, 0, 0, 0.1), 0px 2px 10px rgba(0, 0, 0, 0.1),
    0px 10px 20px rgba(0, 0, 0, 0.1), 0px 10px 50px rgba(0, 0, 0, 0.1);
`;

export const Option = styled.div`
  display: flex;
  align-items: center;
  gap: 1em;
  padding: 0.5em 0.6em;
  border-bottom: 1px solid ${({ theme }) => theme.palette.secondary};
  &:hover {
    transition: background-color 100ms;
    background-color: ${({ theme }) => darken(0.05, theme.palette.accent)};
  }
`;
