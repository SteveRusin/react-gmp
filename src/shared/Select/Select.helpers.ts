import { SelectPropsMultiple } from './Select.models';

export function isMultiSelect<TId>(props: any): props is SelectPropsMultiple<TId> {
  return (props as SelectPropsMultiple<TId>).multi;
}
