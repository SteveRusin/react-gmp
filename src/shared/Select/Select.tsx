import { PureComponent } from 'react';
import { darken } from 'polished';

import { TriangleDown } from '@styled-icons/octicons';
import styled, { ThemeConsumer } from 'styled-components';

import { SelectProps } from './Select.models';
import { SelectOption } from '@shared';

const SelectTriangle = styled(TriangleDown)`
  color: ${(props) => props.color};
  margin-left: auto;
`;

const SelectWrapper = styled.div`
  display: flex;
  align-items: center;
  max-width: 190px;
  min-width: 150px;

  cursor: pointer;
  position: relative;

  .select-placeholder {
    opacity: 0.5;
  }

  .placeholder-wrapper {
    display: flex;
    width: 100%;
  }
`;

const OptionsWrapper = styled.div`
  position: absolute;
  width: 100%;
  left: 0;
  bottom: 0;
  transform: translateY(100%);
  z-index: 1;
  background-color: ${({ theme }) => theme.palette.secondaryVariant};
`;

const Option = styled.div`
  padding: 0.5em 0.6em;
  border-bottom: 1px solid ${({ theme }) => theme.palette.accent};
  &:hover {
    transition: background-color 100ms;
    background-color: ${({ theme }) =>
      darken(0.1, theme.palette.secondaryVariant)};
  }
`;

export class Select<TId> extends PureComponent<SelectProps<TId>> {
  static defaultProps: Partial<SelectProps<unknown>> = {
    placeholder: 'Select something',
  };

  selectClass = 'select-component';

  state = {
    isOpen: false,
  };

  toggleIsOpen = () => {
    this.setState({
      ...this.state,
      isOpen: !this.state.isOpen,
    });
  };

  handleOutsideClick = (e: MouseEvent) => {
    if (!this.state.isOpen) {
      return;
    }

    const target = e.target as HTMLElement;

    const clickedOnTheSelect = target.closest(`.${this.selectClass}`);

    if (clickedOnTheSelect) {
      return;
    }

    this.toggleIsOpen();
  };

  componentDidMount() {
    document.addEventListener('mousedown', this.handleOutsideClick);
  }

  componentWillUnmount() {
    document.removeEventListener('mousedown', this.handleOutsideClick);
  }

  private findSelectedOption() {
    const { value } = this.props;

    if (value == null) {
      return null;
    }

    return this.props.options.find(({ id }) => id === value);
  }

  render() {
    const { isOpen } = this.state;
    const { options, onSelect, placeholder } = this.props;

    const selectedOption = this.findSelectedOption();

    return (
      <SelectWrapper onClick={this.toggleIsOpen} className={this.selectClass}>
        <div className='placeholder-wrapper'>
          <span className='select-placeholder'>
            {selectedOption?.name ?? placeholder}
          </span>
          <ThemeConsumer>
            {(theme) => (
              <SelectTriangle color={theme.palette.primary} size='20' />
            )}
          </ThemeConsumer>
        </div>
        {isOpen && (
          <OptionsWrapper>
            {options.map(({ name, id }: SelectOption<TId>) => (
              <Option
                key={`${id}`}
                onClick={() => onSelect(id)}
                data-value={id}
              >
                {name}
              </Option>
            ))}
          </OptionsWrapper>
        )}
      </SelectWrapper>
    );
  }
}
