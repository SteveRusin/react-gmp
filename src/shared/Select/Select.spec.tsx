import { ThemeProvider } from 'styled-components';
import { defaultTheme } from '@theme';
import userEvent from '@testing-library/user-event';
import { render, screen } from '@testing-library/react';

import { Select } from './Select';
import { SelectProps } from './Select.models';
import { SelectOption } from '@shared';

function renderComponent<T>(props: SelectProps<T>) {
  return (
    <ThemeProvider theme={defaultTheme}>
      <Select<T> {...props} />
    </ThemeProvider>
  );
}

describe('Select Component', () => {
  it('should renders correctly', () => {
    const { asFragment } = render(
      renderComponent({
        onSelect: jest.fn(),
        options: [
          {
            id: 1,
            name: 'Test item',
          },
        ],
      })
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it('should select multiple items when multi=true', () => {
    const FIRST_OPTION = {
      id: 1,
      name: '1',
    };
    const SECOND_OPTION = {
      id: 2,
      name: '2',
    };
    const options: SelectOption<number>[] = [FIRST_OPTION, SECOND_OPTION];

    let selectedValues: number[] = [];
    const onSelect = jest.fn((values) => {
      selectedValues = values;

      rerender(
        renderComponent<number>({
          multi: true,
          options,
          onSelect,
          value: selectedValues,
        })
      );
    });
    const { rerender } = render(
      renderComponent<number>({
        multi: true,
        options,
        onSelect,
        value: selectedValues,
      })
    );

    userEvent.click(screen.getByTestId('select'));
    userEvent.click(screen.getByText(FIRST_OPTION.name), {
      cancelable: true,
    });

    userEvent.click(screen.getByText(SECOND_OPTION.name), {
      cancelable: true,
    });

    expect(selectedValues).toEqual([FIRST_OPTION.id, SECOND_OPTION.id]);
  });
});
