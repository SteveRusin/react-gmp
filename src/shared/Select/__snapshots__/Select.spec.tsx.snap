// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Select Component should renders correctly 1`] = `
<DocumentFragment>
  <div
    class="sc-eCstlR wvyOx select-component"
    data-testid="select"
  >
    <div
      class="placeholder-wrapper"
    >
      <input
        class="sc-gsTEea eECbOD select-trigger"
        placeholder="Select something"
        readonly=""
        value=""
      />
      <svg
        aria-hidden="true"
        class="StyledIconBase-ea9ulj-0 iuNRof sc-hKgJUU dUbcdB"
        color="#F65261"
        fill="currentColor"
        focusable="false"
        height="25"
        viewBox="0 0 16 16"
        width="25"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="m4.427 7.427 3.396 3.396a.25.25 0 0 0 .354 0l3.396-3.396A.25.25 0 0 0 11.396 7H4.604a.25.25 0 0 0-.177.427z"
        />
      </svg>
    </div>
  </div>
</DocumentFragment>
`;
